### PR TITLE
fix(connlib): filter DNS servers on unsupported network stack

### PR DESCRIPTION
**Pull request by thomaseizinger**
Building on top of #6382, we can use the status of our sockets to pro-actively tell our state machine about our network stack. For example, if we don't have an IPv6 socket (anymore), we filter out any IPv6 DNS servers and emit a new event to tell the client to update the available servers.

If we have a broken socket, our interaction with the relays when booting up should already trigger the deactivation of the socket. Thus, we should settle on the right DNS servers before a user starts to access resources.

This surfaces an interesting edge-case: If an admin configures an upstream DNS and a client doesn't have a compatible IP stack, this will result in the client having **no** DNS servers at all. We could also fallback to the system resolvers in that case but we would effectively undermine the admin's decision to force a certain DNS server onto the clients.

Resolves: #6353.

### CODE REVIEW

**Review by jamilbk — APPROVED** [review]
Curious to test this.

**Comment by thomaseizinger** [comment]
> Curious to test this.

Still working on fixing the tests but the actual functionality should hold I think.

**Comment by thomaseizinger** [comment]
> This surfaces an interesting edge-case: If an admin configures an upstream DNS and a client doesn't have a compatible IP stack, this will result in the client having **no** DNS servers at all. We could also fallback to the system resolvers in that case but we would effectively undermine the admin's decision to force a certain DNS server onto the clients.

@jamilbk I've made a call here to honor the upstream settings. Let me know if you'd rather want us to fallback to the system DNS in case the ones in the portal are not compatible.

**Comment by jamilbk** [comment]
> > This surfaces an interesting edge-case: If an admin configures an upstream DNS and a client doesn't have a compatible IP stack, this will result in the client having **no** DNS servers at all. We could also fallback to the system resolvers in that case but we would effectively undermine the admin's decision to force a certain DNS server onto the clients.
> 
> @jamilbk I've made a call here to honor the upstream settings. Let me know if you'd rather want us to fallback to the system DNS in case the ones in the portal are not compatible.

This is fine. Along with #5115 I think we should hopefully avoid that case.

File: rust/connlib/tunnel/src/lib.rs [hunk 1]
```diff
@@ -111,6 +117,8 @@ impl ClientTunnel {
 
     pub fn poll_next_event(&mut self, cx: &mut Context<'_>) -> Poll<Result<ClientEvent, Error>> {
         for _ in 0..MAX_EVENTLOOP_ITERS {
+            self.role_state.set_ip_stack(self.io.ip_stack()?);
```
Comment (conectado) [inline rust/connlib/tunnel/src/lib.rs]: This can only change after sending right? can we set it there?

**Review by conectado — APPROVED** [review]
LGTM

File: rust/connlib/tunnel/src/lib.rs (hunk 1 above)
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/lib.rs]: After sending and after roaming. I am a bit worried that setting it conditionally on something might make us miss a spot later.

Are you worried about performance?

**Comment by thomaseizinger** [comment]
Perhaps instead of reacting to the network stack here, we should only tolerate X number of failures from a certain DNS server before we remove it?
